### PR TITLE
Auto update for debian / rhel platforms & refactoring to support omnitruck

### PR DESCRIPTION
**Pull request by sersut**
```ruby
# SPEC

# Installs specific version of a product. If an older version exists upgrades,
# if a newer version exists downgrades.
chef_ingredient "chef" do
  action :install
  version '12.5.0'
  channel :current # default is :stable
end

# Installs the latest version of a product. If an older or newer version exists,
# does not do anything.
chef_ingredient 'chef' do
  action :install
  channel :stable
end

# Always installs the latest version of a product.
chef_ingredient 'chef' do
  action :upgrade
  channel :current
end

# chef run is always stopped when 'chef' product is installed or upgraded.
# using :reconfigure action with client products results in a warning message
#   and no action.
# chef-ingredient supports installing and upgrading chef on all platforms.
# on platforms that is supported via native repositories, chef-ingredient uses
#   native package managers. Otherwise it uses mixlib-install to install these
#   products from omnitruck.
# all installation specific logic goes into mixlib-install and all resource /
#   provider logic goes into chef-ingredient.
```

Supporting versions other than `:latest` (as explained in https://github.com/chef-cookbooks/chef-ingredient/issues/39) is not in the scope of this work.

There are two remaining items but will have different PRs for them since this is already a monster PR:

- Implement channel pinning on RHEL.
- Enable installations from omnitruck.

### CODE REVIEW

**Comment by thommay** [comment]
this sort of feels like a duplicate of work that's going to need to go into mixlib-install... at the very minimum we've discussed making artifactory/omnitruck/packagecloud supported in m-i

**Comment by lamont-granquist** [comment]
This at least has the right 'shape' in terms of the class layout (composition >> inheritance).  I didn't have time to extensively code review it...

@thommay we probably want to ship this then pull features into mixlib-shellout, then refactor here, iteratively.

**Comment by sersut** [comment]
@lamont-granquist yeah that's what I was thinking. Improving the tests right now. Probably we will merge this without getting the omnitruck things working since omnitruck is currently under development to get support to understand `current` bucket.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+language: ruby
+cache: bundler
+bundler_args: "--jobs 7"
+
+rvm:
+  - 2.1.5
```
Comment (jtimberman) [inline .travis.yml]: worth perhaps using the same version in ChefDK 0.8.0?

File: Gemfile [hunk 2]
```diff
@@ -30,4 +26,7 @@ group :development do
   gem 'guard-rspec'
   gem 'guard-rubocop'
   gem 'mixlib-versioning'
+  gem 'mixlib-install', github: 'chef/mixlib-install'
```
Comment (jtimberman) [inline Gemfile]: Is there not a version released on rubygems.org to use? Should we pin this to a specific shasum?

File: README.md [hunk 3]
```diff
@@ -116,6 +117,30 @@ end
...
+To install or upgrade lastest version of Chef Client on your nodes:
+
+```ruby
+chef_ingredient "chef" do
+  action :upgrade
+  version :latest
+end
+```
+
+To install an addon of Chef Server from `:current` branch:
```
Comment (jtimberman) [inline README.md]: `s/branch/channel/` ;)

File: libraries/chef_ingredient_provider.rb [hunk 4]
```diff
@@ -16,58 +16,82 @@
...
+require_relative './debian_handler'
+require_relative './rhel_handler'
+require_relative './omnitruck_handler'
 
 class Chef
   class Provider
     class ChefIngredient < Chef::Provider::LWRPBase
       provides :chef_ingredient
-      # for include_recipe
+      use_inline_resources
```
Comment (jtimberman) [inline libraries/chef_ingredient_provider.rb]: I believe this is on by default in Chef 12.5.0, fwiw.

File: libraries/chef_ingredient_provider.rb [hunk 5]
```diff
@@ -16,58 +16,82 @@
...
-        execute "#{ingredient_package_name}-reconfigure" do
-          command "#{ctl_command} reconfigure"
+        if ctl_command.nil?
+          Chef::Log.warn "Product '#{new_resource.product_name}' does not support reconfigure."
+          Chef::Log.warn 'chef_ingredient is skipping :reconfigure.'
+        else
+          # Render the config incase it is not rendered yet
+          ingredient_config new_resource.package_name do
+            action :render
+          end
```
Comment (jtimberman) [inline libraries/chef_ingredient_provider.rb]: Oh good call! :icecream:

File: libraries/chef_ingredient_resource.rb [hunk 6]
```diff
@@ -21,22 +21,22 @@ class ChefIngredient < Chef::Resource::LWRPBase
...
-      state_attrs :installed
 
       attribute :product_name, kind_of: String, name_attribute: true
-      attribute :installed, kind_of: [TrueClass, FalseClass, NilClass], default: false
-      attribute :reconfigure, kind_of: [TrueClass, FalseClass], default: false
       attribute :config, kind_of: String, default: nil
 
+      # Attributes for determining what version to install from which channel
+      attribute :version, kind_of: [String, Symbol], default: :latest
+      attribute :channel, kind_of: Symbol, default: :stable, equal_to: [:current, :stable]
```
Comment (jtimberman) [inline libraries/chef_ingredient_resource.rb]: Hmm, perhaps we could allow the strings and then `#to_sym` them?

File: libraries/debian_handler.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+          # If the user specifies "0.0.0", :latest or "latest" as version,
+          # we should not give any version to the package resource.
+          if Mixlib::Versioning.parse(version_string(new_resource.version)) > '0.0.0'
+            version version_for_package_resource
+          end
+
+          if new_resource.product_name == 'chef'
+            notifies :run, 'ruby_block[stop chef run]', :immediately
+          end
```
Comment (jtimberman) [inline libraries/debian_handler.rb]: Should we use `declare_chef_run_stop_resource` here, instead of in the action methods in the provider? And in the redhat_handler, omnitruck_handler?

File: libraries/matchers.rb [hunk 8]
```diff
@@ -70,13 +70,4 @@ def graceful_kill_omnibus_service(pkg)
...
     ChefSpec::Matchers::ResourceMatcher.new(:omnibus_service, :once, pkg)
   end
-
-  def render_ingredient_config(pkg)
-    ChefSpec::Matchers::ResourceMatcher.new(:ingredient_config, :render, pkg)
-  end
-
-  def add_ingredient_config(pkg)
-    ChefSpec::Matchers::ResourceMatcher.new(:ingredient_config, :add, pkg)
-  end
```
Comment (jtimberman) [inline libraries/matchers.rb]: Why are these removed? While we're not using them in the specs, since you used `:step_into`, someone who wants to use these in their own specs will need matchers.

File: libraries/omnitruck_helpers.rb [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    include Helpers
+
+    # TODO: This method should get product name as a parameter and find the
+    # latest version for it.
+    def current_version(product_name)
+      unless %w(chef chefdk).include?(product_name)
+        fail "Unknown product #{product_name}"
+      end
+
+      JSON.parse("/opt/#{product_name}/version-manifest.json")['build_version']
```
Comment (jtimberman) [inline libraries/omnitruck_helpers.rb]: Are the products that have `version-manifest.json` consistently installed in `/opt/PRODUCT_NAME` now? I know we used to have `/opt/opscode` for Enterprise Chef and Private Chef.

File: libraries/omnitruck_helpers.rb [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        action :run
+        code lazy {
+          require 'mixlib/install'
+          installer = Mixlib::Install.new(project: new_resource.product_name, version: version)
+          installer.install_command
+        }
+        if new_resource.product_name == 'chef'
+          notifies :run, 'ruby_block[stop chef run]', :immediately
+        end
+      end
```
Comment (jtimberman) [inline libraries/omnitruck_helpers.rb]: Why a `script` block, and not a `ruby_block` here? Also, wouldn't this need an interpreter?

File: libraries/omnitruck_helpers.rb [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+      script "uninstall-#{new_resource.product_name}" do
+        action :run
+        code lazy {
+          require 'mixlib/install'
+          installer = Mixlib::Install.new(project: new_resource.product_name)
+          installer.uninstall_command
+        }
+      end
```
Comment (jtimberman) [inline libraries/omnitruck_helpers.rb]: Same as above - why `script`?

File: libraries/omnitruck_helpers.rb [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        p: 'mac_os_x',
+        pv: '10.10',
+        m: 'x86_64'
+      }
+    end
+
+    # Returns the channel parameters that omnitruck understands
+    # TODO: This is not quite right. Currently omnitruck does not look at
+    # builds that are posted to s3://opscode-omnibus-packages-current
+    # so we need to change this.
```
Comment (jtimberman) [inline libraries/omnitruck_helpers.rb]: I believe we had a deployment of omnitruck earlier today that fixed that.

File: libraries/rhel_handler.rb [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        rpm_package new_resource.product_name do
+          action action_name
+          package_name ingredient_package_name
+          options new_resource.options
+          source new_resource.package_source
+
+          if new_resource.product_name == 'chef'
+            notifies :run, 'ruby_block[stop chef run]', :immediately
+          end
+        end
```
Comment (jtimberman) [inline libraries/rhel_handler.rb]: I believe `yum_package` can also use a local file system source location. So we could probably use `package`.

Comment (jtimberman) [inline libraries/rhel_handler.rb]: That is we may be able to get away with simplifying this with a single package resource :).

File: libraries/rhel_handler.rb [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+      else
+        # This is to cleanup old cruft from chef-server-ingredient
+        file '/etc/yum.repos.d/chef_stable_.repo' do
+          action :delete
+          only_if { ::File.exist?('/etc/apt/sources.list.d/chef_stable_.list') }
```
Comment (jtimberman) [inline libraries/rhel_handler.rb]: Need RHEL location not Ubuntu's :8ball:.

File: libraries/rhel_handler.rb [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        end
+
+        # Enable the required yum-repository.
+        include_recipe "yum-chef::#{new_resource.channel}"
+
+        # Foodcritic doesn't like timeout attribute in package resource
+        package new_resource.product_name do # ~FC009
+          action action_name
+          package_name ingredient_package_name
+          options new_resource.options
```
Comment (jtimberman) [inline libraries/rhel_handler.rb]: Don't we need the `--enable-repo` and `--disable-repo` options handling here for RHEL like we do the pinning on Ubuntu?

File: libraries/chef_ingredient_provider.rb (hunk 4 above)
Comment (lamont-granquist) [inline libraries/chef_ingredient_provider.rb]: Only if you're using the new-style providers embedded in the resources which this isn't....

File: .travis.yml (hunk 1 above)
Comment (sersut) [inline .travis.yml]: The reason I have selected `2.1.5` is that travis does not have `2.1.6` natively installed yet. Using `2.1.6` makes Travis install Ruby everytime it is running the tests and lengthens the tests by 1 - 2 minutes.

File: Gemfile (hunk 2 above)
Comment (sersut) [inline Gemfile]: :+1:

File: libraries/chef_ingredient_provider.rb (hunk 4 above)
Comment (sersut) [inline libraries/chef_ingredient_provider.rb]: I will leave this in until we move to the new resource writing style.

File: libraries/chef_ingredient_provider.rb (hunk 5 above)
Comment (sersut) [inline libraries/chef_ingredient_provider.rb]: :+1: :smile:

File: libraries/chef_ingredient_resource.rb (hunk 6 above)
Comment (sersut) [inline libraries/chef_ingredient_resource.rb]: IMO unless there is a reason we should not relax this in order not to sprinkle repo with `to_s`s. But if that is a good cookbook writing practice we can do that :smile:

File: libraries/debian_handler.rb (hunk 7 above)
Comment (sersut) [inline libraries/debian_handler.rb]: I moved that to the base class in order not to repeat the call to it. This looks a bit magical here though. I will add a comment here to remind of that.

File: libraries/matchers.rb (hunk 8 above)
Comment (sersut) [inline libraries/matchers.rb]: Ah good call. Will add them back.

File: libraries/omnitruck_helpers.rb (hunk 9 above)
Comment (sersut) [inline libraries/omnitruck_helpers.rb]: That is still the case @jtimberman:

https://github.com/chef/omnibus-software/blob/master/config/software/version-manifest.rb#L23
https://github.com/chef/chef-server/blob/master/omnibus/config/projects/chef-server.rb

Since we are using this class only for chef & chefdk and since omnitruck logic is not enabled yet I think this is fine for now. But will add a reminder comment to fix that later.

File: libraries/omnitruck_helpers.rb (hunk 10 above)
Comment (sersut) [inline libraries/omnitruck_helpers.rb]: Yeah there is a bit more work to do here. E.g. on windows we should select `:powershell`. I would like to fix these details in a separate PR since this is already very big.

File: libraries/rhel_handler.rb (hunk 13 above)
Comment (sersut) [inline libraries/rhel_handler.rb]: So I moved over the original resource structure. I recall that we had some issues with the package provider / resolver and introduced this logic. Do we want to switch it back?

File: libraries/rhel_handler.rb (hunk 14 above)
Comment (sersut) [inline libraries/rhel_handler.rb]: Nice catch :) Full rhel support (similar to the `apt_preference` logic is on its way after this PR).

File: libraries/rhel_handler.rb (hunk 15 above)
Comment (sersut) [inline libraries/rhel_handler.rb]: Not sure if `--enable-repo` and `--disable-repo` is the right ones but yes we need something :smile: Want to add them with a different PR.

File: .travis.yml (hunk 1 above)
Comment (lamont-granquist) [inline .travis.yml]: You should probably use 2.1 and 2.2 though and let it float, similarly to the chef project, minor version pins usually wind up testing against old versions.

**Comment by sersut** [comment]
@jtimberman muchos gracias for the comments. Caught important things. I would like to add the below two things with separate PRs if that's OK with you since this PR is already very big. 

- [ ] Implement channel pinning on RHEL.
- [ ] Enable installations from omnitruck.

File: libraries/chef_ingredient_resource.rb (hunk 6 above)
Comment (lamont-granquist) [inline libraries/chef_ingredient_resource.rb]: Yeah, I'm kind of hugely :-1: on that kind of sugar based on the abuse of attributes and other areas.  There was some `#to_sym` stuff in the old `provides` which I turned into a deprecation warning in Chef 12.0, it just generates more testing and ambiguity and we should just define which the interface takes and force the user to use the one we pick.

File: libraries/chef_ingredient_provider.rb (hunk 4 above)
Comment (jtimberman) [inline libraries/chef_ingredient_provider.rb]: :rocket: :cake:

File: libraries/chef_ingredient_resource.rb (hunk 6 above)
Comment (jtimberman) [inline libraries/chef_ingredient_resource.rb]: Fair points! :rocket: :cake:

File: libraries/omnitruck_helpers.rb (hunk 9 above)
Comment (jtimberman) [inline libraries/omnitruck_helpers.rb]: :rocket: :cake:

File: libraries/omnitruck_helpers.rb (hunk 10 above)
Comment (jtimberman) [inline libraries/omnitruck_helpers.rb]: That's fine then. mixlib/install should get the right platform specific bits so we don't have to think about it.

File: libraries/rhel_handler.rb (hunk 13 above)
Comment (jtimberman) [inline libraries/rhel_handler.rb]: It's "probably fine" but the duplication means that modifying this code may lead to drift if we don't keep everything in line. Hopefully this is fairly complete at this point, though.

File: libraries/rhel_handler.rb (hunk 14 above)
Comment (jtimberman) [inline libraries/rhel_handler.rb]: :tada:

File: libraries/rhel_handler.rb (hunk 15 above)
Comment (jtimberman) [inline libraries/rhel_handler.rb]: :+1:

File: .travis.yml (hunk 1 above)
Comment (jtimberman) [inline .travis.yml]: I think this is fine for now and we can optimize versions later.

**Comment by jtimberman** [comment]
Still an open question about the omnitruck `#channel_parameters` TODO comment, but other than that, I'm :+1: :rocket: :cake:

File: libraries/rhel_handler.rb (hunk 13 above)
Comment (sersut) [inline libraries/rhel_handler.rb]: Yeah I think this is good for now. We can revisit in the future if needed.

File: libraries/omnitruck_helpers.rb (hunk 12 above)
Comment (sersut) [inline libraries/omnitruck_helpers.rb]: Yeah will add a separate PR for it after testing against the new omnitrucks.
